File: chrome/test/data/nacl/ppapi/ppp_instance/ppapi_ppp_instance.js
// Copyright 2013 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

function startsWith(str, prefix) {
  return (str.indexOf(prefix) === 0);
}

function setupTests(tester, plugin) {
  //////////////////////////////////////////////////////////////////////////////
  // Test Helpers
  //////////////////////////////////////////////////////////////////////////////
  var numMessages = 0;
  function addTestListeners(numListeners, test, testFunction, runCheck) {
    var messageListener = test.wrap(function(message) {
      if (!startsWith(message.data, testFunction)) return;
      test.log(message.data);
      numMessages++;
      plugin.removeEventListener('message', messageListener, false);
      test.assertEqual(message.data, testFunction + ':PASSED');
      if (runCheck) test.assert(runCheck());
      if (numMessages < numListeners) {
        plugin.addEventListener('message', messageListener, false);
      } else {
        numMessages = 0;
        test.pass();
      }
    });
    plugin.addEventListener('message', messageListener, false);
  }

  function addTestListener(test, testFunction, runCheck) {
    return addTestListeners(1, test, testFunction, runCheck);
  }

  //////////////////////////////////////////////////////////////////////////////
  // Tests
  //////////////////////////////////////////////////////////////////////////////

  tester.addTest('PPP_Instance::DidCreate', function() {
    assertEqual(plugin.lastError, '');
  });

  tester.addAsyncTest('PPP_Instance::DidChangeView', function(test) {
    // The .cc file hardcodes an expected 15x20 size.
    plugin.width = 15;
    plugin.height = 20;
    addTestListener(test, 'DidChangeView');
  });

  // This test does not appear to be reliable on the bots.
  // http://crbug.com/329511
  /*
  tester.addAsyncTest('PPP_Instance::DidChangeFocus', function(test) {
    // TODO(polina): How can I simulate focusing on Windows?
    // For now just pass explicitely.
    if (startsWith(navigator.platform, 'Win')) {
      test.log('skipping test on ' + navigator.platform);
      test.pass();
      return;
    }
    addTestListeners(2, test, 'DidChangeFocus');
    plugin.tabIndex = 0;
    plugin.focus();
    plugin.blur();
  });
  */

  // PPP_Instance::HandleDocumentLoad is only used with full-frame plugins.
  // This is tested in tests/ppapi_browser/extension_mime_handler/

  // PPP_Instance::DidDestroy is never invoked in the untrusted code.
  // We could wait for a crash event from it, but CallOnMainThread semantics
  // on shutdown are still buggy, so it might never come even if the function
  // triggered. Plus waiting for something not to happen makes the test flaky.
}

File: chrome/test/data/nacl/ppapi/ppp_instance/ppapi_ppp_instance.cc
// Copyright 2013 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <stdint.h>
#include <string.h>

#include <vector>

#include "chrome/test/data/nacl/ppapi_test_lib/get_browser_interface.h"
#include "chrome/test/data/nacl/ppapi_test_lib/test_interface.h"
#include "native_client/src/include/nacl_macros.h"
#include "native_client/src/shared/platform/nacl_check.h"
#include "ppapi/c/pp_bool.h"
#include "ppapi/c/pp_errors.h"
#include "ppapi/c/pp_input_event.h"
#include "ppapi/c/ppb_view.h"
#include "ppapi/c/ppp_instance.h"

namespace {

PP_Bool DidCreate(PP_Instance instance,
                  uint32_t argc,
                  const char* argn[],
                  const char* argv[]) {
  printf("--- PPP_Instance::DidCreate\n");
  PP_Bool status = DidCreateDefault(instance, argc, argn, argv);
  // TEST_PASSED has no effect from this function.
  // But as long as the plugin loads and tests are run, we know this succeeded.
  // See ppapi_browser/bad for failing DidCreate.
  return status;
}

// This should never be called.
void DidDestroy(PP_Instance instance) {
  printf("--- PPP_Instance::DidDestroy\n");
  CHECK(instance == pp_instance());
  NACL_NOTREACHED();
}

void DidChangeView(PP_Instance instance, PP_Resource view) {
  printf("--- PPP_Instance::DidChangeView\n");
  EXPECT(instance == pp_instance());

  PP_Rect clip;
  PPBView()->GetClipRect(view, &clip);
  EXPECT(clip.point.x == 0 && clip.point.y == 0);

  // These are based on embed dimensions.
  PP_Rect position;
  PPBView()->GetRect(view, &position);
  EXPECT(position.size.width == 15 && clip.size.width == 15);
  EXPECT(position.size.height == 20 && clip.size.height == 20);

  TEST_PASSED;
}

void DidChangeFocus(PP_Instance instance,
                    PP_Bool has_focus) {
  printf("--- PPP_Instance::DidChangeFocus has_focus=%d\n", has_focus);
  // There should be no focus on load, so this will trigger when we gain it
  // and then release it and so on.
  static bool expected_has_focus = true;
  EXPECT(instance == pp_instance());
  EXPECT(has_focus == expected_has_focus);
  expected_has_focus = !expected_has_focus;

  TEST_PASSED;
}

PP_Bool HandleDocumentLoad(PP_Instance instance,
                           PP_Resource url_loader) {
  // Only called for full-frame plugins. For coverage see:
  // tests/ppapi_browser/extension_mime_handler/
  NACL_NOTREACHED();
}

const PPP_Instance ppp_instance_interface = {
  DidCreate,
  DidDestroy,
  DidChangeView,
  DidChangeFocus,
  HandleDocumentLoad
};

}  // namespace


void SetupTests() {
  // Each PPP_Instance function called by the browser acts as a test.
}

void SetupPluginInterfaces() {
  RegisterPluginInterface(PPP_INSTANCE_INTERFACE, &ppp_instance_interface);
}
